sapi/fpm/fpm/fpm_stdio.c

```diff
@@ -162,49 +162,53 @@ static void fpm_stdio_child_said(struct fpm_event_s *ev, short which, void *arg)
 		}
 	}
 
-	while (fifo_in || fifo_out) {
-		if (fifo_in) {
-			res = read(fd, buf + in_buf, max_buf_size - 1 - in_buf);
-			if (res <= 0) { /* no data */
-				fifo_in = 0;
-				if (res == 0 || (errno != EAGAIN && errno != EWOULDBLOCK)) {
-					/* pipe is closed or error */
-					read_fail = (res < 0) ? res : 1;
-				}
-			} else {
-				in_buf += res;
-				/* if buffer ends with \0, then the stream will be finished */
-				if (!buf[in_buf - 1]) {
-					finish_log_stream = 1;
-					in_buf--;
-				}
+	while (1) {
+stdio_read:
+		in_buf = read(fd, buf, max_buf_size - 1);
+		if (in_buf <= 0) { /* no data */
+			if (in_buf == 0 || (errno != EAGAIN && errno != EWOULDBLOCK)) {
+				/* pipe is closed or error */
+				read_fail = (in_buf < 0) ? in_buf : 1;
 			}
+			break;
 		}
-
-		if (fifo_out) {
-			if (in_buf == 0) {
-				fifo_out = 0;
+		start = 0;
+		if (cmd_pos > 0) {
+			if 	((sizeof(FPM_STDIO_CMD_FLUSH) - cmd_pos) <= in_buf &&
```

Comment:
I think there is still a potential problem here, if you get something like `\0f` then `sc` then `f`. This would end up flushing the `\0fsc` to the stream right now, if I read things correctly. Not sure if that can happen in practice though.

Reply:
@nikic Yeah I have considered this case before. You are right in theory but this case can't really happen IMHO. The thing is that `\0fscf\0` is written in singe `write` and I think that no supported OS will chunk it in such way that it would be in multiple chunks. So the only thing that can happen is that it is split at the end of line - it can't really be split more once though.

Do you still think that it's an issue?

Reply:
No, I think it's fine. This code is already complicated enough...